### PR TITLE
ENH: stats: add `method` parameter to `differential_entropy` + more accurate estimators for small samples

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-13631

#### What does this implement/fix?
This adds a `method` parameter to `scipy.stats.differential_entropy` with three new estimation methods, all of which tend to be more accurate than the Vasicek estimator for small samples. 

Question: should Vasicek remain the default since it's the best known and just as accurate for large samples, or is small sample accuracy more important?

#### Additional information
See [here](https://link.springer.com/article/10.1007/s40745-015-0045-9) for a reference that compares the methods. It also proposes new methods that may be marginally better for small samples, but I would leave those to a separate PR.

I couldn't find other software that implements these methods, but they're all reasonably well-cited (corresponding papers have 100-200 citations), and they do seem to be consistently more accurate than Vasicek for small samples, at least for the distributions tested.

Here's a plot of the root mean squared error (between the estimate and exact entropy) over 10,000 samples from some common distributions as a function of the number of observations in a sample `n`. Window size is `round(sqrt(n))`.

image

image

image

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thanks for the follow up PR. I compared the code with the maths from the paper you linked and LGTM. I am just not sure I understand your pad function completely. But it looks like something that might be used elsewhere. So maybe we should put this in utils actually?

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:92]: That does sound like a good option. One thought is rather than naming a method, we could make it `auto`, giving us flexibility to change the default over time. Also, the few plots suggest that we might make `auto` mean `van es` for very small samples (<10), `ebrahimi` for moderate samples (10~1000), and `vasicek` beyond that.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:98]: I would actually remove it entirely if we change the default. I don't think we need to define differential entropy here, do you?

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:282]: I'm not sure I understand this one. What extra spaces? Is it whitespace inside the function or double blank line after the function? The latter is a PEP8 requirement.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:92]: I like this! But we should be clear then in the doc that the `auto` behaviour could break the compatibility.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:289]: You're right, but the intent was to avoid modification - even refactoring - of Vasicek where possible to make initial review easier. It does this padding, too, but it doesn't use this function. If you'd like me to refactor Vasicek to make it more consistent with the other methods, I'd be happy to, but for now I left the heart of it alone.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:98]: Either fixing or removing is fine for me. In a lot of functions we don't have maths. I am not sure if we want to move away from the maths in general or not though.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:282]: Oops, I understand now after reading [this](https://github.com/scipy/scipy/pull/13845/files#r611473529). Which would you prefer? I suppose i could do:
```suggestion
    difference = X[..., m:] - X[..., :-m]
    term1 = 1/(n-m) * np.sum(np.log((n+1)/m * difference), axis=-1)
    k = np.arange(m, n+1)
    return term1 + np.sum(1/k) + np.log(m) - np.log(n+1)
```

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:291]: See suggestion [here](https://github.com/scipy/scipy/pull/13845/files#r611777013). Would that address this?

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:296]: It may or may not be faster, but yes, it is the most common case, so your understanding is correct. Mostly I did it to eliminate one line from the logic. I would use a function `twos_like` if I could; this is the closest thing to that.

File: scipy/stats/_entropy.py
Comment (mdhaber) [inline scipy/stats/_entropy.py:315]:
```suggestion
    difference = X[..., j0] - Xibar
    num = np.sum(difference*(j-i), axis=-2)
    den = n*np.sum(difference**2, axis=-2)
```

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:316]: I can see some PEP8 whitespace corrections to be made here, but not necessarily those recommended. PEP8 recommends:
> If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator

So I don't think I would add the whitespace around `**`, for instance. The whitespace around single operators inside functions (e.g. `np.arange(1, n+1)` and `np.log(num / den)`) is optional, and I prefer to omit because it is the highest-priority operation (e.g. add before creating the array, divide before taking the log). On the other hand, when operators are not inside functions, I add the whitepace (e.g. `j = i + dj`) because it looks a bit better - in fact, that's given as a "correct" example in the PEP8 guide.

One thing that is more difficult to justify is my space in `X[..., j0] - Xibar`. I suppose I do that because the indexing is like a high priority operator. (For example, if it were `X**2 - Xibar`, I would want the space, even though it appears inside a function. But if we refactor as [above](https://github.com/scipy/scipy/pull/13845/files#r611782368), then it wouldn't matter anymore.

File: scipy/stats/tests/test_entropy.py
Comment (mdhaber) [reply scipy/stats/tests/test_entropy.py:1]: Oops copy pasted.

File: scipy/stats/tests/test_entropy.py
Comment (mdhaber) [reply scipy/stats/tests/test_entropy.py:5]: Others have argued that `assert` is preferred over `assert_`. (see [here](https://github.com/scipy/scipy/issues/8583)). I think the issue was closed because we do not need to _change_ old uses of `assert_`, but since it _might_ have a slight edge and looks better, might as well use `assert` in new code.

NumPy recommends `assert_allcose` over `assert_almost_equal` and `assert_array_almost_equal`. (see [here](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html).)

File: scipy/stats/tests/test_entropy.py
Comment (mdhaber) [reply scipy/stats/tests/test_entropy.py:249]: I was too! Seems common enough to justify the trivial implementation... but I don't really want to add it.

File: scipy/stats/_entropy.py
Comment (mdhaber) [inline scipy/stats/_entropy.py:200]: @tupui There's no reason to divide all the terms individually;  might as well take the mean first, then divide a single number, right? Also, moving it out of this Vasicek-specific function allows the change of base to be shared by all the methods.

Is that right? Is conversion of entropy from one base ("units") to another as simple as that?

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:265]: I can commit these changes all together when we decide on some others, e.g. [here](https://github.com/scipy/scipy/pull/13845/files#r611782368).

**Comment by mdhaber** [comment]
Thanks for the review @tupui. I think I responded to all the comments except for some simple suggestions that we can commit along with the rest once we reach consensus.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:282]: I wanted to add extra space between the operations so `/ - +`. Here I am not sure we need the variable `difference`. Below I was actually thinking that it might be slower to have this extra variable.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:289]: The refactoring would be ok for me. But you can leave it for now if you want someone else to review after we iterate.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:296]: Ok thanks for confirming this.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:316]: OK sure it makes sense 👍

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:98]: I would love to include the math, formatted as math. However, [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) disagrees, suggesting that LaTeX should only be used in the Notes section, and even then:
image
Rather than risking additional rebuke for my use of LaTeX, I would prefer to remove it entirely if it's not needed.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:282]: Referring to something that has to be calculated takes ~0s extra time.
```
> x = np.random.rand(100, 100, 100, 100)
> %timeit difference = x - x
205 ms ± 1.21 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
> %timeit x - x
208 ms ± 5.75 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:289]: I don't think a second review is needed. If you're willing to merge after reviewing, let me know and I'll refactor Vasicek.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py]: OK, done in 9efd2c626d5fd717a229b320e68128ec76acdf67

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:285]: I went ahead and did it in ebe65b316487dbbbfec22d386b0f9b64a1b6dbb2.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:98]: You have more experience with `scipy.stats`, I will let you judge and remove the maths if not needed :)

Just to be sure, I am talking about the maths in general so this and the rest with the main equation (`\frac{1}{n}\sum_1^n ...`)

**Comment by mdhaber** [comment]
Re: CI errors, I will fix the refguide_asv_check by specifying `method='vasicek'`. I should probably provide more examples anyway.

I'll fix mypy's astute observation `scipy/stats/_entropy.py:243: error: "None" not callable` by replacing `None` with `_vasicek_estimator`. I suppose then I can get rid of the `else` in the auto method selection.

File: scipy/stats/_entropy.py
Comment (mdhaber) [inline scipy/stats/_entropy.py:229]:
```suggestion
               "auto": _vasicek_entropy}  # only used if N > 1000
```

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:200]: Yes makes sense to do it afterwards.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:285]: Looks good thanks. It's more readable and consistent with the rest.

File: scipy/stats/tests/test_entropy.py
Comment (tupui) [reply scipy/stats/tests/test_entropy.py:4]: OK thanks for the references.

File: scipy/stats/_entropy.py
Comment (tupui) [inline scipy/stats/_entropy.py:241]: To be sure, we do this because it's faster than ebrahimi right?

**Comment by tupui** [comment]
> @tupui The original example said:
> 
> ```
>     >>> # We use high numbered seeds in order to have good entropy in the RNG
>     >>> rng = np.random.default_rng(seed=148632)
> ```
> 
> Is it really preferred to use large seeds?

Yes apparently it is better for the generator. But for the examples we should actually remove the seeding at all as we did in #13778. For the result we should then put `# random` so that the doctests do not fail. This is to not advertise any seed pinning.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:229]: Maybe you could define it like `methods: Dict[str, Optional[callable]] = ...` or just skip mypy check here. Does not look good to put a wrong thing here.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:241]: That wasn't my motivation. Vasicek is really not much faster; the only difference is that Ebrahimi weights the `difference` terms at the left/right edges of the sample less strongly. But from the slopes of the lines above, it looks to me like Vasicek will win for larger samples, at least for two of the three distributions.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:229]: It's not necessarily wrong. For instance, we could remove the 
```
        else:
            method = 'vasicek'
```
below and then it would actually be used as part of the logic. If n > 1000, `method` would remaing `auto` and then `methods[method]` would return `_vasicek_entropy`. The would be equivalent.

That said, I would prefer to leave the `else` condition.

Re defining `methods` differently - could you click the `+` mark where that line should be inserted, and click the 
image

button in the comment window to add a specific suggestion to the code?

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py:241]: Ok thanks for the precision's. I agree about this fallback, makes sense.

File: scipy/stats/_entropy.py
Comment (tupui) [inline scipy/stats/_entropy.py:229]: I am not 100% sure this will work due to multiple lines.
This looks better than adding the type as I suggested I think. We would need to import `Dict` just for that otherwise.
```suggestion
               "auto": None}  # type: ignore
```

Comment (tupui) [reply scipy/stats/_entropy.py:229]: I would also keep the `else` to keep it clear. It's obfuscated otherwise.

**Comment by tupui** [comment]
To summarize, IMO this just needs:

* fix of the type hint
* removing of seed in example (we can also do it in the follow up)
* remove or change the equation

**Comment by mdhaber** [comment]
> removing of seed in example (we can also do it in the follow up)

gh-13778 was mostly about removing `RandomState`, but why does it also remove the seed? Why is it better to have results that are not repeatable? Sometimes docstests are not tested in the test suite separately (perhaps they _should_ be, but that's not always the case), and they have helped me find bugs in the past.

**Comment by tupui** [comment]
> > removing of seed in example (we can also do it in the follow up)
> 
> [gh-13778](https://github.com/scipy/scipy/pull/13778) was mostly about removing `RandomState` from `default_rng`, but why does it also remove the seed? Why is it better to have results that are not repeatable? Sometimes docstests are not tested in the test suite separately (perhaps they _should_ be, but that's not always the case), and they have helped me find bugs in the past.

In the end we also discussed about removing the seed in the examples because users tend to copy paste this code and not touch it when it works for them. So this prevents bad practices and common seeds again. It's not a random number generator anymore if you always use the same seed ;).

I also found bugs in `stats.qmc` with doctests, but in the end it was because I was missing a proper test. Our test suite should be more covering than the docstrings. They are the real tests. Still we should continue to run the doctests just to verify that our examples are coherent with the actual implementation.

File: scipy/stats/_entropy.py
Comment (zoj613) [inline scipy/stats/_entropy.py:233]: I don't think you need to be explicit with the `.key()` method here.

File: scipy/stats/_entropy.py
Comment (zoj613) [inline scipy/stats/_entropy.py:165]: Is there a reason why you didn't default to `None` for the method parameter? That way you can say `"if method is not specified, then ....". This has the advantage that "auto" does not appear in the exception message when a user passes the wrong string since it is the default option.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:233]: You're right. I've written this sort of code several times and never before have I been explicit about `keys`! Not sure what was different in my brain that day : )

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:165]: Considered `None`, but I guess I went with how methods are selected in some existing statistical tests, e.g. [`kstest`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kstest.html) and some new ones being added in 1.7 (e.g. [`pagel`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.page_trend_test.html), [`cramervonmises_2samp`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.cramervonmises_2samp.html)). I'll think about it some more, thanks!

File: scipy/stats/_entropy.py
Comment (zoj613) [inline scipy/stats/_entropy.py:306]: Just for my own personal learning here: Is the spacing difference around operators outside and inside function calls intentional? If so, whats the motivation behind it?

Comment (mdhaber) [reply scipy/stats/_entropy.py:306]: [PEP8 whitespace recommendations](https://www.python.org/dev/peps/pep-0008/#other-recommendations) allows some flexibility, so some of it is personal judgement of what makes sense.

Comment (zoj613) [reply scipy/stats/_entropy.py:306]: Mmm, I didn't know this. Thanks for the reference.

File: scipy/stats/_entropy.py
Comment (mdhaber) [inline scipy/stats/_entropy.py:232]:
```suggestion
    if method not in methods:
        message = f"`method` must be one of {set(methods)}"
```

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:165]: I've thought about it a bit and if we didn't want it to appear in the exception message, I can see the advantage. But I don't really see why it shouldn't appear in the exception message. Either `None` or `auto` is a valid value of the parameter.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:233]: Pushing fix soon.

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py:98]: @tupui [This](https://github.com/scipy/scipy/pull/13298#pullrequestreview-638207522) is a perfect example of why I now avoid including math unless necessary. Please briefly note the amount of work that has gone into that PR. In that case, the comment is about the notes section, where LaTeX is allowed, and it is a statistical distribution, where LaTeX is always used to define the distribution. So I would prefer to just remove the LaTeX here.

**Comment by mdhaber** [comment]
@tupui I think I addressed the comments, except I'll let you remove the seed in gh-13863. Could you check the doc build locally? I added another example and I want to make sure it looks alright. I'll try to check, too, but I haven't successfully built docs locally in over a year.

**Comment by tupui** [comment]
> @tupui I think I addressed the comments, except I'll let you remove the seed in [gh-13863](https://github.com/scipy/scipy/pull/13863). Could you check the doc build locally? I added another example and I want to make sure it looks alright. I'll try to check, too, but I haven't successfully built docs locally in over a year.

Sorry @mdhaber for the delay, I oversaw this! I could test your branch locally, the doctests are passing and the doc is building just fine. (You might want to merge master again to have the doc fix).
Just one thing, there is not link to references in the description of the methods.


Screenshot 2021-04-20 at 13 25 32

Screenshot 2021-04-20 at 13 24 55
Screenshot 2021-04-20 at 13 25 08

File: scipy/stats/_entropy.py
Comment (tupui) [inline scipy/stats/_entropy.py:159]:
```suggestion
    * ``'vasicek'`` uses the estimator presented in [1]_. This is
      one of the first and most influential estimators of differential entropy.
    * ``'van es'`` uses the bias-corrected estimator presented in [3]_, which is
      not only consistent but, under some conditions, asymptotically normal.
    * ``'ebrahimi'`` uses an estimator presented in [4]_, which was shown
      in simulation to have smaller bias and mean squared error than
      the Vasicek estimator.
    * ``'correa'`` uses the estimator presented in [5]_ based on local linear
      regression. In a simulation study, it had consistently smaller mean
      square error than the Vasiceck estimator, but it is more expensive to
      compute.
    * ``'auto'`` selects the method automatically (default). Currently,
      this selects ``'van es'`` for very small samples (<10), ``'ebrahimi'``
      for moderate sample sizes (11-1000), and ``'vasicek'`` for larger
      samples, but this behavior is subject to change in future versions.

    All estimators are implemented as described in [6]_.
```

**Comment by mdhaber** [comment]
Thanks @tupui I guess to get the numbers right in the docstring I'll just have to wait to see the values CI reports and change them to that. I have so far been unsuccessful at predicting the values because doctests for stats fail on my machine (memory error). I also tried the seed you told me, but I don't think that got me the right answer either. (I will try that again though. It's possible I did something out of order.)

**Comment by mdhaber** [comment]
Looks like the Linux Tests / Nightly Cpython (3.9) failure is unrelated and happening in other PRs (e.g. gh-13906). I'm happy with this if you are!

File: scipy/stats/_entropy.py
Comment (mdhaber) [inline scipy/stats/_entropy.py:119]: Actually do you know how to properly continue this line? It is showing up like:

image

Comment (mdhaber) [reply scipy/stats/_entropy.py:119]: If we decide what to do here, I can remove the "We use high numbered seeds in order to have good entropy in the RNG" below at the same time.

Comment (zoj613) [reply scipy/stats/_entropy.py:119]: A work around would be to drop the "str in" since the set items already imply `str` type. I noticed that is what is done in many docstrings of scipy functions. Maybe even drop the space in "van es". Maybe this way optional might fit in the previous line.

Comment (mdhaber) [reply scipy/stats/_entropy.py]: Thanks! That did it.

image

Comment (tupui) [reply scipy/stats/_entropy.py]: Ah sorry I missed this! There is a way, you just need to skip a blank line before the description. It's what I had to do for the `seed`:

```python
    seed : {None, int, `numpy.random.Generator`,
            `numpy.random.RandomState`}, optional

        If `seed` is None the `numpy.random.Generator` singleton is used.
        If `seed` is an int, a new ``Generator`` instance is used,
        seeded with `seed`.
        If `seed` is already a ``Generator`` or ``RandomState`` instance then
        that instance is used.
```

**Comment by tupui** [comment]
> @tupui Looks like the Linux Tests / Nightly Cpython (3.9) failure is unrelated and happening in other PRs (e.g. [gh-13906](https://github.com/scipy/scipy/pull/13906)). I'm happy with this if you are! ~_Update: oops, some little things noted below._~

LGTM too! The doctests where passing locally, so you did it correctly with the seed.

**Review by tupui — APPROVED** [review]

File: scipy/stats/_entropy.py
Comment (mdhaber) [reply scipy/stats/_entropy.py]: Yeah I found that later tonight, too. But it's fine I think.

**Comment by mdhaber** [comment]
> The doctests where passing locally,

I already fixed it I guess.

File: scipy/stats/_entropy.py
Comment (tupui) [reply scipy/stats/_entropy.py]: Sure, sure 👍

File: scipy/stats/_entropy.py
Comment (tupui) [inline scipy/stats/_entropy.py:215]: Leftover. Will not impact any test as for plotting only.
```suggestion
    ...        rvs = stats.expon.rvs(size=(reps, n), random_state=rng)
```

**Comment by tupui** [comment]
I will just wait for `refguide_asv_check` and merge as the rest was green before.

**Comment by tupui** [comment]
Thanks @mdhaber!

**Pull request merged**
